services/api/service.go

```diff
@@ -1013,18 +991,50 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 		BlockNumber: payload.ExecutionPayload.BlockNumber,
 		NumTx:       uint64(len(payload.ExecutionPayload.Transactions)),
 	}
-	err = api.datastore.SaveBid(&bidTrace, &getHeaderResponse, &getPayloadResponse)
+
+	//
+	// Save to Redis
+	//
+	// first the trace
+	err = api.redis.SaveBidTrace(&bidTrace)
 	if err != nil {
-		log.WithError(err).Error("could not save bid and block")
-		api.RespondError(w, http.StatusBadRequest, err.Error())
+		log.WithError(err).Error("failed saving bidTrace in redis")
+		api.RespondError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// save execution payload (getPayload response)
+	err = api.redis.SaveExecutionPayload(payload.Message.Slot, payload.Message.ProposerPubkey.String(), payload.Message.BlockHash.String(), &getPayloadResponse)
+	if err != nil {
+		log.WithError(err).Error("failed saving execution payload in redis")
+		api.RespondError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// save this builder's latest bid
+	err = api.redis.SaveLatestBuilderBid(payload.Message.Slot, payload.Message.BuilderPubkey.String(), payload.Message.ParentHash.String(), payload.Message.ProposerPubkey.String(), &getHeaderResponse)
+	if err != nil {
+		log.WithError(err).Error("could not save latest builder bid")
+		api.RespondError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// recalculate top bid
+	err = api.redis.UpdateTopBid(payload.Message.Slot, payload.Message.ParentHash.String(), payload.Message.ProposerPubkey.String())
```

Comment:
Hm, you only need to recalculate top bid if the overwritten bid was the top one right?
Would it be worth to check?

Reply:
seems cheaper to just write, rather than reading the top bid from redis on every submission as well? 🤔